### PR TITLE
feat(core_ext presence): add `.present?` to check for data

### PR DESCRIPTION
**Pull request by stakach**
simplifies two cases:
* checking if a variable is not `nil?` and not `empty?`
* assigning a default value `optional = param.present? || ["default"]`

### CODE REVIEW

**Comment by kimburgess** [comment]
Given the similarity, to remove potential confusion for people coming from rails, is it worth mirroring [their API](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/object/blank.rb) for this?

This way we could have:
```crystal
present? : Bool
```
```crystal
presence : self?
```
Then also add a `presence(&block : self -> T) : T? forall T`, which provides something that feel like `Object#try` but for ‘blank-ness’.

**Comment by kimburgess** [comment]
Actually, just checked the latest stdlib and it has this already implemented for both `String` and `Nil` already too.

**Comment by stakach** [comment]
oh cool! Going to start using `presence` everywhere.

**Comment by stakach** [comment]
adding this back as `presence` is only implemented for `Nil` and `Strings` in the standard lib

File: src/core_ext/presence.cr [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class Object
+  # returns its receiver if it's not nil? or empty?
+  def present?
```
Comment (caspiano) [inline src/core_ext/presence.cr:3]:
```suggestion
  def presence
```

File: spec/core_ext/presence_spec.cr [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+             when 2
+               [1, 2, 3, 4]
+             when 3
+               {1, 2, 3}
+             when 4
+               {"me" => 123}
+             end
+
+    typeof(object).to_s.should eq("(Array(Int32) | Hash(String, Int32) | String | Tuple(Int32, Int32, Int32) | Nil)")
+    (!!object.present?).should eq(true)
```
Comment (caspiano) [inline spec/core_ext/presence_spec.cr:36]:
```suggestion
    (!!object.presence).should eq(true)
```

File: spec/core_ext/presence_spec.cr [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+describe "core_ext/presence" do
+  it "checks for presence of data in objects" do
+    nil.present?.should eq nil
+    Array(String).new.present?.should eq nil
+    Hash(String, String).new.present?.should eq nil
+
+    1234.present?.should eq 1234
+    [1, 2, 3].present?.should eq [1, 2, 3]
+    {"test" => "data"}.present?.should eq({"test" => "data"})
```
Comment (caspiano) [inline spec/core_ext/presence_spec.cr:12]:
```suggestion
    nil.presence.should eq nil
    Array(String).new.presence.should eq nil
    Hash(String, String).new.presence.should eq nil

    1234.presence.should eq 1234
    [1, 2, 3].presence.should eq [1, 2, 3]
    {"test" => "data"}.presence.should eq({"test" => "data"})
```

File: spec/core_ext/presence_spec.cr [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  it "checks whitespace in strings" do
+    "".present?.should eq nil
+    "  ".present?.should eq nil
+    "   \t\n  \n".present?.should eq nil
+    "  data ".present?.should eq "  data "
```
Comment (caspiano) [inline spec/core_ext/presence_spec.cr:19]:
```suggestion
    "".presence.should eq nil
    "  ".presence.should eq nil
    "   \t\n  \n".presence.should eq nil
    "  data ".presence.should eq "  data "
```

File: src/core_ext/presence.cr [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    end
+  end
+end
+
+# a special case to return `nil` if a String only contains whitespace
+class String
+  def present?
+    self if !blank?
+  end
+end
```
Comment (caspiano) [inline src/core_ext/presence.cr:18]: `String#presence` is already defined in the standard library

```suggestion
```

**Review by caspiano — COMMENTED** [review]
I think `presence` might be a more suitable name as the same method is already implemented in the stdlib for `String`.

**Comment by stakach** [comment]
Started a conversation on the crystal forums @KimBurgess @caspiano would be pretty simple pull request
https://forum.crystal-lang.org/t/object-presence/2355

**Comment by stakach** [comment]
refactored to be more inline with the std lib

**Pull request closed without merging**